components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -4570,4 +4565,281 @@ public String getLifecycleConfiguration(String tenantDomain) throws APIManagemen
+                        apiProduct.getId().getApiProductName() + ", APIProduct Version : " +
+                        apiProduct.getId().getVersion());
+            }
+        } catch (APIManagementException e) {
+            throw new APIManagementException("Error in adding API :" + apiProduct.getId().getApiProductName(), e);
+        }
+    }
+
+    /**
+     * Create an Api
+     *
+     * @param apiProduct API
+     * @throws APIManagementException if failed to create API
+     */
+    private void createAPIProduct(APIProduct apiProduct) throws APIManagementException {
+        GenericArtifactManager artifactManager = APIUtil.getArtifactManager(registry, APIConstants.API_PRODUCT_KEY);
+
+        boolean transactionCommitted = false;
+        try {
+            registry.beginTransaction();
+            GenericArtifact genericArtifact =
+                    artifactManager.newGovernanceArtifact(new QName(apiProduct.getId().getApiProductName()));
+            GenericArtifact artifact = APIUtil.createAPIProductArtifactContent(genericArtifact, apiProduct);
+            artifactManager.addGenericArtifact(artifact);
+            //Attach the API lifecycle
+            artifact.attachLifecycle(APIConstants.API_LIFE_CYCLE);
+            String artifactPath = GovernanceUtils.getArtifactPath(registry, artifact.getId());
+            String providerPath = APIUtil.getAPIProductProviderPath(apiProduct.getId());
+            //provider ------provides----> API
+            registry.addAssociation(providerPath, artifactPath, APIConstants.PROVIDER_ASSOCIATION);
+            Set<String> tagSet = apiProduct.getTags();
+            if (tagSet != null) {
+                for (String tag : tagSet) {
+                    registry.applyTag(artifactPath, tag);
+                }
+            }
+
+            //write Status to a separate property. This is done to support querying APIProducts using custom query (SQL)
+            //to gain performance
+            String apiStatus = apiProduct.getStatus().getStatus();
+            saveAPIStatus(artifactPath, apiStatus);
+            String visibleRolesList = apiProduct.getVisibleRoles();
+            String[] visibleRoles = new String[0];
+            if (visibleRolesList != null) {
+                visibleRoles = visibleRolesList.split(",");
+            }
+            APIUtil.setResourcePermissions(apiProduct.getId().getProviderName(), apiProduct.getVisibility(),
+                    visibleRoles, artifactPath);
+            registry.commitTransaction();
+            transactionCommitted = true;
+
+            if (log.isDebugEnabled()) {
+                String logMessage =
+                        "API Name: " + apiProduct.getId().getApiProductName() + ", APIProduct Version "
+                                + apiProduct.getId().getVersion() + " created";
+                log.debug(logMessage);
+            }
+        } catch (Exception e) {
+            try {
+                registry.rollbackTransaction();
```

Comment:
Why do u need to do registry.rollbackTransaction(); in the catch block ? It seems you have done it in the finally block.